### PR TITLE
AMQP-637: Fix Shutdown

### PR DESCRIPTION
**Pull request by garyrussell**
JIRA: https://jira.spring.io/browse/AMQP-637

### CODE REVIEW

**Comment by garyrussell** [comment]
Don't merge; needs polish

File: spring-rabbit/src/main/java/org/springframework/amqp/rabbit/listener/DirectMessageListenerContainer.java [hunk 1]
```diff
@@ -387,16 +407,36 @@ public void handleDelivery(String consumerTag, Envelope envelope,
...
+		}
+
+		@Override
 		public void handleCancelOk(String consumerTag) {
+			if (this.logger.isDebugEnabled()) {
+				this.logger.debug("CancelOk " + this);
+			}
+			synchronized (DirectMessageListenerContainer.this.consumersMonitor) {
+				DirectMessageListenerContainer.this.cancellationLock.release(this);
+				DirectMessageListenerContainer.this.consumers.remove(this);
```
Comment (artembilan) [inline spring-rabbit/src/main/java/org/springframework/amqp/rabbit/listener/DirectMessageListenerContainer.java]: Why don't we remove `Consumer` from the `consumersByQueue` as well, like we do that in the `handleCancel()`?

Comment (garyrussell) [inline spring-rabbit/src/main/java/org/springframework/amqp/rabbit/listener/DirectMessageListenerContainer.java]: I guess we could for consistency - right now it's removed in adjustConsumers when we cancel (or cleared in doShutdown).

Comment (artembilan) [inline spring-rabbit/src/main/java/org/springframework/amqp/rabbit/listener/DirectMessageListenerContainer.java]: OK. Merging with that fix.

Comment (garyrussell) [inline spring-rabbit/src/main/java/org/springframework/amqp/rabbit/listener/DirectMessageListenerContainer.java]: No; I pushed

**Comment by artembilan** [comment]
OK. Merging...

**Comment by artembilan** [comment]
Merged as https://github.com/spring-projects/spring-amqp/commit/1834b6c9f4fafdec5038316f2deea9e163b99863

**Pull request closed without merging**